#231 First step in refactoring clustering towards shared use with ALA

timrobertson100 opened the pull request:
This is the first step in refactoring the code to be shared between GBIF and ALA.

This change adds an interface for accessing the OccurrenceFeatures, and an implementation using of this against a Spark Row (as it was).

A Pojo implementation is added in the Test folder (using a verbose builder to avoid bring the new style to this project). The tests are repeated on both the Spark version (as they originally were) and the Pojo version.

My intention is to release this in occurrence and then move the code to pipelines and remove it from this project. I wish to split the refactor and the move into two steps to avoid changing too many things in one go. 

Note: The code will be reformated to the google code style during moving to pipelines, and the use of lombok will be added in that move.

@MattBlissett - happy to take a call to orient you on the changes before you review.

timrobertson100 commented on occurrence-clustering/src/main/java/org/gbif/occurrence/clustering/OccurrenceRelationships.java:
```diff
@@ -176,27 +178,22 @@ private static void compareIdentifiers(OccurrenceFeatures o1, OccurrenceFeatures
   }
 
-  static boolean presentOnOneOnly(OccurrenceFeatures o1, OccurrenceFeatures o2, String field) {
-    return (o1.get(field) == null && o1.get(field) != null) || (o1.get(field) != null && o1.get(field) == null);
```
This was seemingly wrong before the rector (all `o1` and no `o2`) and fixed here.

MattBlissett commented on occurrence-clustering/src/main/java/org/gbif/occurrence/clustering/RowOccurrenceFeatures.java:
```diff
@@ -0,0 +1,200 @@
+
+  <T> T get(String field) {
+    int fieldIndex = assertNotNull(row.fieldIndex(prefix + field), "Field not found in row schema: "
+      + prefix + field);
```
`row.fieldIndex` return an int (not null), and throws an exception if the field doesn't exist.

MattBlissett commented:
Other than that, this looks fine to me.  We probably want the list of "No aplica" etc to be configurable and not duplicated, but I doubt there's any point implementing that before it's moved into the pipelines project.

I committed fixes to some minor typos etc myself.

timrobertson100 commented:
@MattBlissett I addressed feedback, squashed to a single commit, and rebased pulling in the Junit5 changes.
Can you cast a quick eye over that it looks as you'd expect and merge, please?
